local-cli/server/util/attachHMRServer.js

```diff
@@ -16,11 +16,254 @@ const url = require('url');
+                  dependencies: modulesToUpdate,
+                });
+              });
+
+          });
+        })
+        .then((resolutionResponses) => {
+          if (clients.length === 0 || !resolutionResponses) {
+            return;
+          }
+
+          return forEachBundleEntry((bundleEntry, platform) => {
+            if (!resolutionResponses[platform] || !resolutionResponses[platform][bundleEntry]) {
+              return;
+            }
+
+            if (!platformCache[platform][bundleEntry].shallowDependencies[filename]){
+              return;
+            }
+
+            const resolutionResponse = resolutionResponses[platform][bundleEntry];
+
+            const httpServerAddress = httpServer.address();
+
+            // Sanitize the value from the HTTP server
+            let packagerHost = 'localhost';
+            if (httpServerAddress.address &&
+                httpServerAddress.address !== '::' &&
+                httpServerAddress.address !== '') {
+              packagerHost = httpServerAddress.address;
+            }
+
+            let packagerPort = httpServerAddress.port;
+            return packagerServer.buildBundleForHMR({
+              entryFile: bundleEntry,
+              platform,
+              resolutionResponse,
+            }, packagerHost, packagerPort);
+          });
+        })
+        .then(bundles => {
+          if (clients.length === 0 || !bundles) {
+            return;
+          }
+
+          return forEachBundleEntry((bundleEntry, platform) => {
+            if (!bundles[platform]) {
+              return;
+            }
+
+            const bundle = bundles[platform][bundleEntry];
+
+            if (!bundle || bundle.isEmpty()) {
+              return;
+            }
+            return {
+              type: 'update',
+              body: {
+                modules: bundle.getModulesCode(),
+                sourceURLs: bundle.getSourceURLs(),
+                sourceMappingURLs: bundle.getSourceMappingURLs(),
+              },
+            };
+          });
+        })
+        .catch(error => {
+          // send errors to the client instead of killing packager server
+          let body;
+          if (error.type === 'TransformError' ||
+              error.type === 'NotFoundError' ||
+              error.type === 'UnableToResolveError') {
+            body = {
+              type: error.type,
+              description: error.description,
+              filename: error.filename,
+              lineNumber: error.lineNumber,
+            };
+          } else {
+            console.error(error.stack || error);
+            body = {
+              type: 'InternalError',
+              description: 'react-packager has encountered an internal error, ' +
+                'please check your terminal error output for more details',
+            };
+          }
+
+          return forEachBundleEntry((bundleEntry, platform) => ({type: 'error', body}));
+        })
+        .then(update => {
```

Comment:
I think you want to do this `then` for each of the clients instead of waiting for all of the to compute their update. The resulting code would be simpler as you'll be able to handle each client individually. Error handling will become easier.

Moving forward: I think we need to model each client with a class and have a function that returns a promise to handle the file change, update the cache if needed, send the HMR update and do error handling. The cache should live inside of this class as it's not really shared across multiple clients. The only case where it could actually be reused is if you're testing something both on the simulator and on the device. On this case we should share the cache but I don't think it's that big of a deal not to do it for now. To add support for this the client class would have to have an array of web sockets to which it should send the updates. I say we separate the PR in 2: one to add support for multiple platforms/apps and other one to support this very special use case. What do you think?